iceoryx2-cal/src/zero_copy_connection/used_chunk_list.rs

```diff
@@ -0,0 +1,205 @@
+            allocator: &T,
+        ) -> Result<(), iceoryx2_bb_elementary::allocator::AllocationError> {
+            if self.is_memory_initialized.load(Ordering::Relaxed) {
+                fatal_panic!(from self,
+                "Memory already initialized. Initializing it twice may lead to undefined behavior.");
+            }
+
+            let memory = fail!(from self, when allocator
+            .allocate(Layout::from_size_align_unchecked(
+                    std::mem::size_of::<AtomicBool>() * self.capacity,
+                    std::mem::align_of::<AtomicBool>())),
+            "Failed to initialize since the allocation of the data memory failed.");
+
+            self.data_ptr.init(memory);
+
+            for i in 0..self.capacity {
+                unsafe {
+                    (self.data_ptr.as_ptr() as *mut AtomicBool)
+                        .add(i)
+                        .write(AtomicBool::new(false))
+                };
+            }
+
+            self.is_memory_initialized.store(true, Ordering::Relaxed);
+
+            Ok(())
+        }
+
+        fn memory_size(capacity: usize) -> usize {
+            Self::const_memory_size(capacity)
+        }
+
+        unsafe fn new(capacity: usize, distance_to_data: isize) -> Self {
+            Self {
+                data_ptr: RelocatablePointer::new(distance_to_data),
+                capacity,
+                is_memory_initialized: AtomicBool::new(true),
+            }
+        }
+    }
+
+    impl<PointerType: PointerTrait<AtomicBool> + Debug> UsedChunkList<PointerType> {
+        pub const fn const_memory_size(capacity: usize) -> usize {
+            std::mem::size_of::<AtomicBool>() * capacity + std::mem::align_of::<AtomicBool>() - 1
+        }
+
+        pub fn capacity(&self) -> usize {
+            self.capacity
+        }
+
+        #[inline(always)]
+        fn verify_init(&self, source: &str) {
+            debug_assert!(
+                self.is_memory_initialized.load(Ordering::Relaxed),
+                "Undefined behavior when calling \"{}\" and the object is not initialized.",
+                source
+            );
+        }
+
+        fn set(&self, idx: usize, value: bool) -> bool {
+            self.verify_init("set");
+            debug_assert!(
+                idx < self.capacity,
+                "This should never happen. Out of bounds access with index {}.",
+                idx
+            );
+
+            unsafe { (*self.data_ptr.as_ptr().add(idx)).swap(value, Ordering::Relaxed) }
+        }
+
+        pub fn insert(&self, value: usize) -> bool {
+            !self.set(value, true)
+        }
+
+        pub fn remove(&self, value: usize) -> bool {
+            self.set(value, false)
+        }
+
+        pub fn pop(&self) -> Option<usize> {
+            self.verify_init("pop");
+
+            (0..self.capacity).find(|&i| unsafe {
+                (*self.data_ptr.as_ptr().add(i)).swap(false, Ordering::Relaxed)
+            })
+        }
```

Comment:
Okay, this is quite expensive when the list is large and the indices at the end are in use. I guess this is only used on cleanup?

Reply:
Exactly, this is only used when a subscriber goes out of scope, terminates the connection and the publisher has to recollect all samples.
I explicitly did not return a `Vec` here to avoid using the heap. The type-state pattern is also out of the question, since this must stick to the same memory position.

But I will improve it a bit by storing the last pop position and reuse it in the next call. When it reached capacity it will start with zero again.

Reply:
No need to complicate things

Reply:
Actually, it makes things easier and more efficient. I will adjust this.